Lib/email/feedparser.py

```diff
@@ -534,3 +559,85 @@ class BytesFeedParser(FeedParser):
 
     def feed(self, data):
         super().feed(data.decode('ascii', 'surrogateescape'))
+
+
+class NonIdentityTransferDecodingFeedParser(abc.ABC):
+    """
+    This is an abstract base class; only its subclasses should be instantiated
+     directly.
+
+    Instances of this class work like FeedParser except that the feed() method,
+     prior to parsing the input, transparently decodes the input consistent with
+     RFC 2045, s. 6.2.  Each concrete subclass reverses one of the non-identity
+     Content-Transfer-Encoding transformations described there.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self._bytes_feed_parser = BytesFeedParser(*args, **kwargs)
+
+    @abc.abstractmethod
+    def feed(self, text):
+        pass
+
+    def close(self):
+        return self._bytes_feed_parser.close()
+
+
+class Base64TransferDecodingFeedParser(NonIdentityTransferDecodingFeedParser):
+    """
+    Concrete, "base64" implementation of TransferDecodingFeedParser.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.encoded_buffer = bytearray()
+
+    def feed(self, line):
+        self.encoded_buffer.extend(line.rstrip().encode('ascii'))
+        total_decodable = int(len(self.encoded_buffer) / 4) * 4
+        if total_decodable >= 1:
+            decodable_bytes = self.encoded_buffer[:total_decodable]
+            self.encoded_buffer = self.encoded_buffer[total_decodable:]
+            decoded_bytes = base64.standard_b64decode(decodable_bytes)
+            self._bytes_feed_parser.feed(decoded_bytes)
+
+    def close(self):
+        if len(self.encoded_buffer) >= 1:
+            decoded_bytes = base64.standard_b64decode(self.encoded_buffer)
+            self._bytes_feed_parser.feed(decoded_bytes)
+            self.encoded_buffer.clear()
+        return self._bytes_feed_parser.close()
+
+
+class QuotedPrintableTransferDecodingFeedParser(
+    NonIdentityTransferDecodingFeedParser
+):
+    """
+    Concrete, "quoted-printable" implementation of TransferDecodingFeedParser.
+    """
+
+    def feed(self, line):
+        if line == '':
+            return
+        if line[-1] != '\n':
+            line += '\r\n'
+        encoded_bytes = line.encode('ascii')
+        decoded_bytes = quopri.decodestring(encoded_bytes)
+        self._bytes_feed_parser.feed(decoded_bytes)
+
+
+def _new_transfer_decoding_parser(content_transfer_encoding, **kwargs):
+    """
+    Creates a new FeedParser object that transparently reverses the specified
+     content_transfer_encoding transformation.
+    :param content_transfer_encoding: str
+    :return: Union[NoneType,NonIdentityTransferDecodingFeedParser] None if
```

Comment:
If you're going to provide this information, use type hints:

```python
def _new_transfer_decoding_parser(content_transfer_encoding: str, **kwargs
    ) -> Union[NoneType, NonIdentityTransferDecodingFeedParser]: 
```
Formatting is tricky because the bass class name is so long. Maybe consider shortening it a bit: `NoIdTransDecodingFP`. I don't really love that abbreviation myself and I like the current descriptive name, but it is quite a mouthful.

Reply:
Thanks, I incorporated the recommendation in a massive update I'm about to apply.